apt2aptly/src/lib.rs

```diff
@@ -0,0 +1,369 @@
+    Ok(dsc_files[0])
+}
+
+#[tracing::instrument(skip(builder, root_location, release))]
+async fn scan_dist_sources(
+    builder: &mut OriginContentBuilder,
+    root_location: &OriginLocation,
+    release: &dyn ReleaseReader,
+    component: &str,
+) -> Result<()> {
+    info!("Scanning sources");
+
+    let entry = release.sources_entry(component)?;
+    let mut reader = entry_reader(release, &entry, entry.compression).await?;
+    while let Some(paragraph) = reader.read_paragraph().await? {
+        let source = DebianSourceControlFile::from(paragraph);
+        let package = source
+            .source()
+            .or_else(|_| source.required_field_str("Package"))
+            .map_err(|_| eyre!("Missing Source/Package field"))?
+            .into();
+
+        let span = info_span!("scan_dist_sources:package", ?package);
+        let _enter = span.enter();
+
+        let files = collect_source_files(&source)?;
+        let dsc = find_dsc_file(&files)?;
+
+        let mut aptly_hash_builder = AptlyHashBuilder::default();
+        for file in &files {
+            aptly_hash_builder.add_file(&AptlyHashFile {
+                basename: &file.name,
+                size: file.size,
+                md5: &file.md5,
+                sha1: &file.sha1,
+                sha256: &file.sha256,
+            });
+        }
+
+        builder.add_dsc(OriginDsc {
+            package,
+            dsc_location: root_location
+                .join(source.required_field_str("Directory")?)?
+                .join(&dsc.name)?,
+            files,
+            aptly_hash: aptly_hash_builder.finish(),
+        });
+    }
+
+    Ok(())
+}
+
+pub type OriginContentByComponent = HashMap<String, OriginContent>;
+
+#[tracing::instrument(fields(root_url = root_url.as_str()), skip(root_url))]
+pub async fn scan_dist(root_url: &Url, dist: &str) -> Result<OriginContentByComponent> {
+    let root_location = OriginLocation::Url(root_url.clone());
+
+    let mut contents = OriginContentByComponent::new();
+
+    let root = HttpRepositoryClient::new(root_url.clone())?;
+    let release = root.release_reader(dist).await?;
+
+    let architectures = release
+        .release_file()
+        .architectures()
+        .ok_or_else(|| eyre!("Release file has no architectures"))?
+        .collect::<Vec<_>>();
+    for component in release
+        .release_file()
+        .components()
+        .ok_or_else(|| eyre!("Release file has no components"))?
+    {
+        let mut builder = OriginContentBuilder::new();
+
+        for arch in &architectures {
+            scan_dist_packages(&mut builder, &root_location, &*release, component, arch).await?;
+        }
+
+        scan_dist_sources(&mut builder, &root_location, &*release, component).await?;
+
+        contents.insert(component.to_owned(), builder.build());
+    }
+
+    Ok(contents)
+}
+
+pub struct BinarySyncer;
+
+#[async_trait::async_trait]
+impl Syncer for BinarySyncer {
```

Comment:
Seems a bit od that the sync crate doesn't implement the actual syncing algorithm ? Why are there seperate ones for both apt2aptly and obs2aptly, syncing is syncing right :)